discovery/tailscale/tailscale.go

```diff
@@ -0,0 +1,243 @@
+package tailscale
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+
+	"tailscale.com/client/tailscale"
+)
+
+const (
+	tsLabel                                 = model.MetaLabelPrefix + "tailscale_"
+	deviceLabel                             = tsLabel + "device_"
+	tsLabelDeviceAddresses                  = deviceLabel + "addresses"
+	tsLabelDeviceID                         = deviceLabel + "id"
+	tsLabelDeviceName                       = deviceLabel + "name"
+	tsLabelDeviceHostname                   = deviceLabel + "hostname"
+	tsLabelDeviceUser                       = deviceLabel + "user"
+	tsLabelDeviceClientVersion              = deviceLabel + "client_version"
+	tsLabelDeviceUpdate                     = deviceLabel + "update"
+	tsLabelDeviceOS                         = deviceLabel + "os"
+	tsLabelDeviceKeyExpiryDisavle           = deviceLabel + "key_expiry_disabled"
+	tsLabelDeviceAuthorized                 = deviceLabel + "authorized"
+	tsLabelDeviceIsExternal                 = deviceLabel + "is_external"
+	tsLabelDeviceMachineKey                 = deviceLabel + "machine_key"
+	tsLabelDeviceNodeKey                    = deviceLabel + "node_key"
+	tsLabelDeviceBlocksIncommingConnections = deviceLabel + "blocks_incomming_connections"
+	tsLabelDeviceEnabledRoutes              = deviceLabel + "enabled_routes"
+	tsLabelDeviceAdvertisedRoutes           = deviceLabel + "advertised_routes"
+	tsLabelDeviceTags                       = deviceLabel + "tags"
+	tsLabelDeviceTailnetLockKey             = deviceLabel + "tailnet_lock_key"
+)
+
+var (
+	DefaultSDConfig = SDConfig{
+		RefreshInterval:  model.Duration(60 * time.Second),
+		Port:             443,
+		URL:              "https://api.tailscale.com",
+		Tailnet:          "-",
+		HTTPClientConfig: config.DefaultHTTPClientConfig,
+		TagSeperator:     ",",
+	}
+	matchContentType = regexp.MustCompile(`^(?i:application\/json(;\s*charset=("utf-8"|utf-8))?)$`)
+	userAgent        = fmt.Sprintf("Prometheus/%s", version.Version)
+	failuresCount    = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "prometheus_sd_tailscale_failures_total",
+			Help: "Number of tailscale service discovery refresh failures.",
+		})
+)
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+	prometheus.MustRegister(failuresCount)
+}
+
+type SDConfig struct {
+	HTTPClientConfig config.HTTPClientConfig `yaml:",inline"`
+	RefreshInterval  model.Duration          `yaml:"refresh_interval,omitempty"`
+	URL              string                  `yaml:"url"`
```

Comment:
```suggestion
	URL              string                  `yaml:"url"`
```

can we use api_url maybe?

Reply:
moved to the `api_url` and `api_token` as well for the API key. Better to stay away from the default HTTP fields (`bearer_token` and `authorization.credentials`). Unless the preference is to migrate to the `authorization` scheme that's on main.